File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/09_10진수를_2진수로_변환하기.js



/**
 * 풀이
 *   10 -> 2진수 로 변환하기
 * 분석
 *   TC: O(N)
 *   - 2진수 변환: O(N)
 *   - 문자열 합치기: O(M)
 *   - O(N + M) -> O(N)
 *   SC: O(N)
 *   - 추가공간 res(int[]) 밖에 없음
 *
 * @param n {number}
 * @returns {string}
 */
function solution(n) {
    const res = []
    while (n !== 1) {
        const left = n % 2
        n = Math.floor(n / 2)
        res.push(String(left))
    }

    res.push('1')
    return res.reduce((a, b) => b + a, "")
}

const r1 = solution(10)
// 10 / 2 = 5 ... 0
// 5  / 2 = 2 ... 1
// 2  / 2 = 1 ... 0
// -> 1010
console.log(r1)

const r2 = solution(27)
// 27 / 2 = 13 ... 1
// 13 / 2 = 6 ... 1
// 6 / 2 = 3 ...0
// 3 / 2 = 1 ... 1
// -> 11011
console.log(r2)

// -> 1100000011100
const r3 = solution(12345)
console.log(r3)

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/09_10진수를_2진수로_변환하기.go
package week02

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func solution09(n int) string {
	str := make([]string, 0)
	num := float64(n)
	for num != 1 {
		left := int64(num) % 2
		num = math.Floor(num / 2)
		str = append(str, strconv.Itoa(int(left)))
	}
	str = append(str, "1")
	return strReverse(strings.Join(str, ""))
}

func strReverse(s string) string {
	rns := []rune(s)
	for i, j := 0, len(rns)-1; i < j; i, j = i+1, j-1 {
		rns[i], rns[j] = rns[j], rns[i]
	}
	return string(rns)
}

func Run09() {
	r1 := solution09(10)
	fmt.Println(r1)

	r2 := solution09(27)
	fmt.Println(r2)

	r3 := solution09(12345)
	fmt.Println(r3)
}
